cmdstanpy/model.py

```diff
@@ -330,29 +340,37 @@ def compile(
                     if self._compiler_options is not None:
                         cmd.extend(self._compiler_options.compose())
                     cmd.append(Path(exe_file).as_posix())
-                    try:
-                        msg = do_command(cmd, cmdstan_path())
-                        if msg is not None and 'Warning or error:' in msg:
-                            msg = msg.split("Warning or error:", 1)[1].strip()
-                            get_logger().warning(
-                                "stanc3 has produced warnings:\n%s", msg
-                            )
 
-                    except RuntimeError as e:
+                    sys_stdout = io.StringIO()
+                    with contextlib.redirect_stdout(sys_stdout):
+                        try:
+                            do_command(cmd=cmd, cwd=cmdstan_path())
+                        except RuntimeError as e:
+                            runtime_error = str(e)
+                            compilation_failed = True
+                        finally:
+                            console = sys_stdout.getvalue()
+
+                    if '--warn-pedantic' in console:  # warn, don't fail
+                        lines = console.split('\n')
+                        warnings = [x for x in lines if x.startswith('Warning')]
+                        get_logger().warning('\n'.join(warnings))
+
+                    if 'Syntax error' in console:
+                        get_logger().warning(console)
```

Comment:
I'm concerned these conditionals are overfitting the warnings stanc3 emits. We emit warnings even without `--warn-pedantic`, and there are compile errors other than syntax-based ones (like semantic errors)

Reply:
is there a way to run the Stanc3 parser to see what error messages are available?
and shouldn't this be documented somewhere - user's guide?

Reply:
Not sure I follow your question entirely, but stanc3 outputs 3 kinds of error intentionally (see [this file](https://github.com/stan-dev/stanc3/blob/master/src/middle/Errors.ml) if you don't mind a little ocaml):

1. Syntax errors from either the lexer or parser, which begin "Syntax error"
2. Semantic (type) errors, which begin "Semantic error"
3. A FileNotFound error (I think we do our own check that the file exists, so this shouldn't happen?)

There are also some  internal compiler errors which should never leak into userspace, but could. They usually begin "This should never happen. Please file a bug"

Reply:
If we want to put it somewhere in the doc, I think the cmdstan guide page on `stanc` would be good, underneath the command line arguments.

Also, for warnings, we mostly emit deprecation warnings if warn-pedantic is not enabled, but there are a few other warnings that get emitted. I know there is a warning about int division with `/` instead of `%/%` that is outside pedantic mode, and another warning for if your program is entirely empty.

Reply:
ran a few of the models in Stanc3 repo dir `test/integration/bad/` - looks like semantic error messages are just like syntactic error messages, only they start with "Semantic error".  will check for this as well.

Reply:
revamped logic, based on current Stanc3 behavoir (stanc3/test/integration/bad)